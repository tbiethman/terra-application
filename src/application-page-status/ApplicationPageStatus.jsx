import { useRef, useContext, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import Button from 'terra-button';
import uuidv4 from 'uuid/v4';
import ApplicationPageStatusContext from './ApplicationPageStatusContext';

const propTypes = {
  /**
   * An array of objects containing terra-button properties. A key attribute is required for each object.
   * This array is used to render buttons in the bottom section.
   * Example:`[{ text: 'Button 1', key: 1, size: 'medium', variant: 'action', onClick: onClickFunction}]`
   */
  // eslint-disable-next-line react/forbid-foreign-prop-types
  buttonAttrs: PropTypes.arrayOf(PropTypes.shape(Button.propTypes)),

  /**
   * The descriptive text, displayed under the title.
   */
  message: PropTypes.string,

  /**
   * The title displayed under the glyph. Variants contain default titles that can be overridden by this prop.
   */
  title: PropTypes.string,

  /**
   * Sets the glyph and title using a pre-baked variant. One of the following: `no-data`,
   * `no-matching-results`, `not-authorized`, or `error`
   */
  variant: PropTypes.oneOf(['no-data', 'no-matching-results', 'not-authorized', 'error']),
};

const defaultProps = {
  buttonAttrs: [],
  message: undefined,
  title: undefined,
  variant: undefined,
};

const ApplicationPageStatus = (props) => {
  const {
    buttonAttrs,
    message,
    title,
    variant,
  } = props;
  const id = useRef(uuidv4());
  const applicationPageStatus = useContext(ApplicationPageStatusContext);

  useLayoutEffect(() => {
    const statusViewId = id.current;

    applicationPageStatus.show(statusViewId, {
      buttonAttrs,
      message,
      title,
      variant,
    });

    return () => {
      applicationPageStatus.hide(statusViewId);
    };
  }, [buttonAttrs, message, title, variant, id, applicationPageStatus]);

  return null;
};

ApplicationPageStatus.propTypes = propTypes;
ApplicationPageStatus.defaultProps = defaultProps;

export default ApplicationPageStatus;
