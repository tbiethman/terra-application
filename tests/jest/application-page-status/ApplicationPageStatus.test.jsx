import React from 'react';
import uuidv4 from 'uuid/v4';

import ApplicationPageStatus from '../../../src/application-page-status/ApplicationPageStatus';
import ApplicationPageStatusContext from '../../../src/application-page-status/ApplicationPageStatusContext';

jest.mock('uuid/v4');

describe('ApplicationPageStatus', () => {
  let reactUseContext;
  let pageStatusContextValue;

  beforeAll(() => {
    uuidv4.mockReturnValue('test-id');

    /**
     * Until Enzyme is updated with full support for hooks, we need to
     * mock out the useContext implementation.
     */
    reactUseContext = React.useContext;
    React.useContext = (contextValue) => {
      if (contextValue === ApplicationPageStatusContext) {
        return pageStatusContextValue;
      }
      return reactUseContext(contextValue);
    };
  });

  afterAll(() => {
    React.useContext = reactUseContext;
  });

  beforeEach(() => {
    pageStatusContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };
  });

  it('should render status view without any data', () => {
    const wrapper = mount(
      <ApplicationPageStatus />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(pageStatusContextValue.show.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();

    expect(pageStatusContextValue.show.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should render status view with the specified data', () => {
    const StatusViewButtons = [
      {
        text: 'Action 1',
        key: 1,
        size: 'medium',
      }, {
        text: 'Action 2',
        key: 2,
        size: 'medium',
      },
    ];

    const wrapper = mount(
      <ApplicationPageStatus
        buttonAttrs={StatusViewButtons}
        message="Status View with all props specified"
        title="Jest Test"
        variant="no-data"
      />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(pageStatusContextValue.show.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();

    expect(pageStatusContextValue.show.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should redisplay status view with new props', () => {
    const wrapper = mount(
      <ApplicationPageStatus message="Error status view" variant="error" />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(pageStatusContextValue.show.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(0);

    wrapper.setProps({ message: 'No data status view', variant: 'no-data' });

    expect(pageStatusContextValue.show.mock.calls.length).toBe(2);
    expect(pageStatusContextValue.show.mock.calls[1][0]).toBe('test-id');
    expect(pageStatusContextValue.hide.mock.calls.length).toBe(1);
    expect(pageStatusContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should honor buttonAttrs prop', () => {
    const StatusViewButtons1 = [
      {
        text: 'Action 1',
        key: 1,
        size: 'medium',
      }, {
        text: 'Action 2',
        key: 2,
        size: 'medium',
      },
    ];

    const StatusViewButtons2 = [
      {
        text: 'Action 3',
        key: 3,
        size: 'medium',
      },
    ];

    const wrapper = mount(
      <ApplicationPageStatus buttonAttrs={StatusViewButtons1} />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ buttonAttrs: StatusViewButtons2 });
    expect(wrapper).toMatchSnapshot();
  });

  it('should honor message prop', () => {
    const wrapper = mount(
      <ApplicationPageStatus message="First message" />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ message: 'Second message' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should honor title prop', () => {
    const wrapper = mount(
      <ApplicationPageStatus title="First title" />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ title: 'Second title' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should honor variant prop', () => {
    const wrapper = mount(
      <ApplicationPageStatus variant="no-data" />,
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'no-matching-results' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'not-authorized' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ variant: 'error' });
    expect(wrapper).toMatchSnapshot();
  });
});
