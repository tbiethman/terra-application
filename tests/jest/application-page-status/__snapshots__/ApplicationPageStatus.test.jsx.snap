// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationPageStatus should honor buttonAttrs prop 1`] = `
<ApplicationPageStatus
  buttonAttrs={
    Array [
      Object {
        "key": 1,
        "size": "medium",
        "text": "Action 1",
      },
      Object {
        "key": 2,
        "size": "medium",
        "text": "Action 2",
      },
    ]
  }
/>
`;

exports[`ApplicationPageStatus should honor buttonAttrs prop 2`] = `
<ApplicationPageStatus
  buttonAttrs={
    Array [
      Object {
        "key": 3,
        "size": "medium",
        "text": "Action 3",
      },
    ]
  }
/>
`;

exports[`ApplicationPageStatus should honor message prop 1`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  message="First message"
/>
`;

exports[`ApplicationPageStatus should honor message prop 2`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  message="Second message"
/>
`;

exports[`ApplicationPageStatus should honor title prop 1`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  title="First title"
/>
`;

exports[`ApplicationPageStatus should honor title prop 2`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  title="Second title"
/>
`;

exports[`ApplicationPageStatus should honor variant prop 1`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  variant="no-data"
/>
`;

exports[`ApplicationPageStatus should honor variant prop 2`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  variant="no-matching-results"
/>
`;

exports[`ApplicationPageStatus should honor variant prop 3`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  variant="not-authorized"
/>
`;

exports[`ApplicationPageStatus should honor variant prop 4`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  variant="error"
/>
`;

exports[`ApplicationPageStatus should redisplay status view with new props 1`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
  message="Error status view"
  variant="error"
/>
`;

exports[`ApplicationPageStatus should render status view with the specified data 1`] = `
<ApplicationPageStatus
  buttonAttrs={
    Array [
      Object {
        "key": 1,
        "size": "medium",
        "text": "Action 1",
      },
      Object {
        "key": 2,
        "size": "medium",
        "text": "Action 2",
      },
    ]
  }
  message="Status View with all props specified"
  title="Jest Test"
  variant="no-data"
/>
`;

exports[`ApplicationPageStatus should render status view without any data 1`] = `
<ApplicationPageStatus
  buttonAttrs={Array []}
/>
`;
